Sources/MapboxCoreNavigation/Router.swift

```diff
@@ -395,16 +395,31 @@ extension InternalRouter where Self: Router {
             guard routeIsFaster else {
                 self.isRerouting = false; return
             }
-            var routeOptions: RouteOptions?
-            if case let .route(options) = response.options {
-                routeOptions = options
-            }
+            
+            let completion = { [weak self] in
+                guard let self = self else { return }
+                var routeOptions: RouteOptions?
+                if case let .route(options) = response.options {
+                    routeOptions = options
+                }
 
-            // Prefer the most optimal route (the first one) over the route that matched the original choice.
-            self.updateRoute(with: indexedResponse,
-                             routeOptions: routeOptions ?? self.routeProgress.routeOptions,
-                             isProactive: true) { [weak self] success in
-                self?.isRerouting = false
+                DispatchQueue.main.async {
+                    // Prefer the most optimal route (the first one) over the route that matched the original choice.
+                    self.updateRoute(with: indexedResponse,
+                                     routeOptions: routeOptions ?? self.routeProgress.routeOptions,
+                                     isProactive: true) { [weak self] success in
+                        self?.isRerouting = false
```

Comment:
I wonder what is the correct way to handle a possible delay before completion call.

 Will reroute work correctly if location changed a lot before the user called completion, e.g. because of some work on the user side in delegate method implementation.

Reply:
It should not be a problem. SDK correctly handles cases when user starts not from the very beginning of the route, so it will just skip already traversed part. The most troublesome case is if completion is called so far away, so that user has already passed the deviation point. In this case, setting this route will result in triggering a re-routing scenario. This is undesired, by currently we don't have good tools to check that, and I consider such case to be very unlikely.